File: 31.07.25 Т-Банк Т-академия/E.py
from collections import deque


def solve(N, segments):
    graph = [[] for _ in range(N + 1)]
    for l, r in segments:
        graph[l - 1].append(r)

    dist = [-1] * (N + 1)
    dist[0] = 0
    queue = deque([0])

    while queue:
        u = queue.popleft()
        for v in graph[u]:
            if dist[v] == -1:
                dist[v] = dist[u] + 1
                queue.append(v)
                if v == N:
                    # нашли путь до N, можно прервать
                    break

    if dist[N] == -1:
        print("No")
    else:
        print("Yes")
        print(dist[N])


# Чтение входных данных
N, Q = map(int, input().split())
segments = [tuple(map(int, input().split())) for _ in range(Q)]

solve(N, segments)

File: 31.07.25 Т-Банк Т-академия/E.java
import java.util.*;

public class E {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int q = scanner.nextInt();

        // Инициализация графа
        List<Integer>[] graph = new ArrayList[n + 1];
        for (int i = 0; i <= n; i++) {
            graph[i] = new ArrayList<>();
        }

        // Построение графа на основе фактов
        for (int i = 0; i < q; i++) {
            int l = scanner.nextInt();
            int r = scanner.nextInt();
            int u = l - 1;
            int v = r;
            graph[u].add(v);
            graph[v].add(u);
        }

        // BFS для поиска кратчайшего пути
        int[] dist = new int[n + 1];
        Arrays.fill(dist, -1);
        Queue<Integer> queue = new LinkedList<>();
        dist[0] = 0;
        queue.add(0);

        while (!queue.isEmpty()) {
            int u = queue.poll();
            for (int v : graph[u]) {
                if (dist[v] == -1) {
                    dist[v] = dist[u] + 1;
                    queue.add(v);
                }
            }
        }

        // Проверка достижимости вершины N
        if (dist[n] == -1) {
            System.out.println("No");
        } else {
            System.out.println("Yes");
            System.out.println(dist[n]);
        }
    }
}